### PR TITLE
Add webpack plugin for dynamic bundling

### PR DESCRIPTION
**Pull request by jkrems**

### CODE REVIEW

File: packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+export default class DynamicBundlePlugin {
+  /**
+   * @param buildBatchEndpoint Helper to construct an asset URL from a serving
+   *                           prefix and a list of chunks ids.
+   */
+  constructor(private buildBatchEndpoint: (prefix: string, ids: number[]) => string) {
+  }
+
+  apply(compiler: webpack.Compiler) {
+    compiler.hooks.compilation.tap(DynamicBundlePlugin.name, (compilation) => {
```
Comment (fenghaolw) [inline packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts]: Add type compilation: webpack.compilation.Compilation?

File: packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        } = compilation.outputOptions;
+
+        const requireFn = '__webpack_require__';
+
+        const scriptNonce = `${requireFn}.nc`;
+        const activeBatch = `${requireFn}._ab`;
+        const publicPath = `${requireFn}.p`;
+
+        return Template.asString([
+          "var script;",
```
Comment (fenghaolw) [inline packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts]: Any reasons to not use ES6 syntax (const/let, arrow function, etc.) in the template?
Will the template be transpiled later or do we need to make it compatible for all browsers?

Seems like you at least use for...of... later so I assume we can use ES6 syntax?

File: packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                    `script.crossOrigin = ${JSON.stringify(crossOriginLoading)};`
+                  ),
+                  "}"
+                ])
+                : "",
+              "// create error before stack unwound to get useful stacktrace later",
+              "var error = new Error();",
+              "onScriptComplete = function (event) {",
+              Template.indent([
+                "// avoid mem leaks in IE.",
```
Comment (fenghaolw) [inline packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts]: So this means IE is not releasing the unreferenced onerror/onload functions?

File: packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+/**
+ * Simulate the chunk loaded callback in webpack 4. Could be dropped if we don't
+ * care about working with webpack 4. This function is stringified and will run
+ * in the browser. `installedChunks` comes from the scope it gets inserted into.
+ */
+function loadingEnded4() {
+  var returnValue = null;
+  // @ts-ignore
```
Comment (fenghaolw) [inline packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts]: Why is this needed? Is that because webpack4 is not well-typed?

File: packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts (hunk 1 above)
Comment (fenghaolw) [inline packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts]: nit: is the styleguide to use (foo) => {} or foo => {}?

Comment (jkrems) [reply packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts]: Can't find a clear note on this in our internal styleguide but once the type annotation is added, the non-parenthesis style is no longer valid syntax.

File: packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts (hunk 4 above)
Comment (jkrems) [reply packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts]: The problem is that this function is serialized and inserted into a different scope. Added a local comment to that effect, more details are in the function-level comment.

File: packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts (hunk 3 above)
Comment (jkrems) [reply packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts]: yeah, IE has (had?) some memory leaks around closures on DOM nodes. This was lifted from the original webpack template, not sure what support level they imply.

File: packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts (hunk 2 above)
Comment (jkrems) [reply packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts]: I think it will generally not be down-leveled since it doesn't count as an input source. Removing the `for...of` below. Nice catch!

File: packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts (hunk 1 above)
Comment (jkrems) [reply packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts]: Done!

**Comment by jkrems** [comment]
PTAL - updated!
